irr/src/CGLTFMeshFileLoader.cpp

```diff
@@ -0,0 +1,610 @@
+*/
+IAnimatedMesh* CGLTFMeshFileLoader::createMesh(io::IReadFile* file)
+{
+	if (file->getSize() <= 0) {
+		return nullptr;
+	}
+	std::optional<tiniergltf::GlTF> model = tryParseGLTF(file);
+	if (!model.has_value()) {
+		return nullptr;
+	}
+
+	if (!(model->buffers.has_value()
+			&& model->bufferViews.has_value()
+			&& model->accessors.has_value()
+			&& model->meshes.has_value()
+			&& model->nodes.has_value())) {
+		return nullptr;
+	}
+
+	ISkinnedMesh *mesh = new CSkinnedMesh();
+	MeshExtractor parser(std::move(model.value()), mesh);
+	try {
+		parser.loadNodes();
+	} catch (std::runtime_error &e) {
+		mesh->drop();
+		return nullptr;
+	}
+	return mesh;
+}
+
+static void transformVertices(std::vector<video::S3DVertex> &vertices, const core::matrix4 &transform)
+{
+	for (auto &vertex : vertices) {
+		// Apply scaling, rotation and rotation (in that order) to the position.
+		transform.transformVect(vertex.Pos);
+		// For the normal, we do not want to apply the translation.
+		// TODO note that this also applies scaling; the Irrlicht method is misnamed.
+		transform.rotateVect(vertex.Normal);
+		// Renormalize (length might have been affected by scaling).
+		vertex.Normal.normalize();
+	}
+}
+
+static void checkIndices(const std::vector<u16> &indices, const std::size_t nVerts)
+{
+	for (u16 index : indices) {
+		if (index >= nVerts)
+			throw std::runtime_error("index out of bounds");
+	}
+}
+
+static std::vector<u16> generateIndices(const std::size_t nVerts)
+{
+	std::vector<u16> indices(nVerts);
+	for (std::size_t i = 0; i < nVerts; i += 3) {
+		// Reverse winding order per triangle
+		indices[i] = i + 2;
+		indices[i + 1] = i + 1;
+		indices[i + 2] = i;
+	}
+	return indices;
+}
+
+/**
+ * Load up the rawest form of the model. The vertex positions and indices.
+ * Documentation: https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#meshes
+ * If material is undefined, then a default material MUST be used.
+*/
+void CGLTFMeshFileLoader::MeshExtractor::loadMesh(
+		const std::size_t meshIdx,
+		ISkinnedMesh::SJoint *parent) const
+{
+	for (std::size_t j = 0; j < getPrimitiveCount(meshIdx); ++j) {
+		auto vertices = getVertices(meshIdx, j);
+		if (!vertices.has_value())
+			continue; // "When positions are not specified, client implementations SHOULD skip primitive’s rendering"
+
+		// Excludes the max value for consistency.
+		if (vertices->size() >= std::numeric_limits<u16>::max())
+			throw std::runtime_error("too many vertices");
+
+		// Apply the global transform along the parent chain.
+		transformVertices(*vertices, parent->GlobalMatrix);
+		
+		auto maybeIndices = getIndices(meshIdx, j);
+		std::vector<u16> indices;
+		if (maybeIndices.has_value()) {
+			indices = std::move(*maybeIndices);
+			checkIndices(indices, vertices->size());
+		} else {
+			// Non-indexed geometry
+			indices = generateIndices(vertices->size());
+		}
+
+		auto *meshbuf = m_irr_model->addMeshBuffer();
+		meshbuf->append(vertices->data(), vertices->size(),
```

Comment:
you can even avoid copying the indices (I think) if you directly touch the relevant variables here.

Reply:
same for the vertices even